blueocean-dashboard/src/main/js/util/smart-fetch.js

```diff
@@ -0,0 +1,277 @@
+                onComplete(data);
+            } catch (e) {
+                // TODO how best to handle failures
+                console.error(e); // eslint-disable-line no-console
+            }
+            return data; // for downstream .then
+        });
+        return holder.promise;
+    }
+}
+
+const duplicateCallTracker = new DuplicateCallTracker();
+
+/**
+ * Mark with $success flag and freeze the object
+ */
+const successAndFreeze = obj => {
+    const out = obj;
+    out.$success = true;
+    deepFreeze(out);
+    return out;
+};
+
+/**
+ * Calls statusCallback with an object holding the current status flag
+ * either .$pending, .$success, or .$failed
+ * Prevents duplicate fetches for the same URL
+ */
+export function fetch(url, options, onData) {
+    let _onData = onData;
+    let _options = options;
+    if (!_onData && typeof _options === 'function') {
+        _onData = _options;
+        _options = null;
+    }
+    if (_onData) {
+        _onData({ $pending: true });
+        return duplicateCallTracker.promise(url, () =>
+            isoFetch(url, _onData || fetchOptions) // Fetch data
+            .then(checkStatus) // Validate success
+            .then(parseJSON) // transfer to json
+            .then(successAndFreeze) // add success field & freeze graph
+            , (data) => {
+                _onData(data);
+            })
+            .catch(err => {
+                _onData({ $failed: err });
+            });
+    }
+    // return a fake promise, a thenable
+    // so it can be resolved multiple times
+    return {
+        then: (fn) => fetch(url, _onData, fn),
+    };
+}
+
+function assignObj(obj, vals) {
+    obj.map = function () {
+        var out = Array.prototype.map.apply(this, arguments);
+        applyFetchMarkers(out, this);
+        return out;
+    };
+    if (vals) {
+        Object.assign(obj, vals);
+    }
+    return obj;
+}
+
+const fetchPagedData = (pager, concatenator, onData, existingData, first, limit) => {
+    // Indicate pending
+    onData(assignObj(concatenator(pager, existingData), { $pending: true, $pager: pager }));
+    return duplicateCallTracker.promise(pager.urlProvider(first, limit + 1), () =>
+        isoFetch(pager.urlProvider(first, limit + 1), fetchOptions) // Fetch data
+        .then(checkStatus) // Validate success
+        .then(parseJSON) // transfer to json
+        .then(successAndFreeze), // add success field & freeze graph
+        data => {
+            // fetched an extra to test if more
+            const hasMore = data.length > limit;
+            const outData = assignObj(concatenator(pager, existingData, data));
+            outData.$success = true;
+            outData.$pager = Object.assign(pager, {
+                current: first,
+                hasMore,
+                currentData: outData,
+                startIndex: existingData.length > 0 ? pager.startIndex : first,
+            });
+            Object.freeze(outData); // already deep frozen
```

Comment:
What does the "already deep frozen" comment here mean? If it's already frozen then why call it again?